ouroboros-network/src/Ouroboros/Network/ErrorPolicy.hs

```diff
@@ -0,0 +1,313 @@
+{-# LANGUAGE DeriveFunctor         #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+
+-- | Error policies, and integration with 'SuspendDecision'-semigroup action on
+-- 'PeerState'.
+--
+module Ouroboros.Network.ErrorPolicy
+  ( ErrorPolicies (..)
+  , ConnectionException (..)
+  , ApplicationException (..)
+  , nullErrorPolicies
+  , ErrorPolicy (..)
+  , evalErrorPolicy
+  , evalErrorPolicies
+  , CompleteApplication
+  , Result (..)
+  , completeApplicationTx
+
+  -- * Traces
+  , ErrorPolicyTrace (..)
+  , traceErrorPolicy
+  , WithAddr (..)
+
+  -- * Re-exports of PeerState
+  , PeerStates
+  , SuspendDecision (..)
+  ) where
+
+import           Control.Exception (Exception, SomeException (..))
+import           Data.List.NonEmpty (NonEmpty (..))
+import           Data.Foldable (traverse_)
+import qualified Data.Map.Strict as Map
+import           Data.Maybe (fromMaybe, mapMaybe)
+import           Data.Semigroup (sconcat)
+import           Data.Set (Set)
+import qualified Data.Set as Set
+import           Data.Typeable ( Proxy (..)
+                               , gcast
+                               , tyConName
+                               , typeRepTyCon
+                               , typeRep
+                               )
+import           Text.Printf
+
+import           Control.Monad.Class.MonadAsync
+import           Control.Monad.Class.MonadSTM
+import           Control.Monad.Class.MonadTime
+import           Control.Tracer
+
+import           Data.Semigroup.Action
+
+import           Ouroboros.Network.Subscription.PeerState
+
+data ErrorPolicy e where
+     ErrorPolicy :: forall e err.
+                      Exception err
+                   => (e err -> SuspendDecision DiffTime)
```

Comment:
Thinking out loud: is this `e` even necessary here? Now when I have to write my `ErrorPolicy`s, I have to deal with this `e` each time, in my case it will always be `ApplicationException`. *Can't this `e` be a phantom parameter?*

Reply:
I don't know if that is possible.

Reply:
I think it can be done with just a phantom parameter. @karknu but if you'll find your self in a corner then we can clean it afterwards.

Reply:
I addressed this in the last commit.